src/authentication/authentication-hapi-plugin.ts

```diff
@@ -38,23 +47,184 @@ class AuthenticationHapiPlugin extends HapiPlugin {
+      const teamToken = await generateAndSaveTeamToken(teamId, this.db);
+      return reply(teamToken).code(201);
+    } catch (error) {
+      return reply(Boom.badRequest(error.message));
+    }
+  }
+
+  public async signupHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    let email: string | undefined;
+    let credentials: AccessToken | undefined;
+    try {
+      credentials = request.auth.credentials as AccessToken;
+      // The email has been validated in validateUser at this point
+      email = credentials[subEmailClaimKey]!;
+      const teamToken = credentials[teamTokenClaimKey]!;
+      const teamId = await getTeamIdWithToken(teamToken, this.db);
+      const team = await this.gitlab.getGroup(teamId);
+      const password = generatePassword();
+      const {id, idp} = parseSub(credentials.sub);
+      const user = await this.gitlab.createUser(
+        email,
+        password,
+        email.replace('@', '-'),
+        email,
+        id,
+        idp,
+      );
+      await this.gitlab.addUserToGroup(user.id, teamId);
+      return reply({
+        team,
+        password,
+      }).code(201); // created
+    } catch (error) {
+      const message = `Problems on signup for user ${email}`;
+      this.logger.error(message, {
+        error,
+        credentials,
+      });
+      return reply(Boom.badRequest(message));
+    }
   }
 
   public async validateUser(
     payload: AccessToken,
-    _request: any,
+    request: any,
     callback: (err: any, valid: boolean, credentials?: any) => void,
   ) {
+
+    let valid = false;
+    let email: string | undefined;
+    try {
+      const credentials = payload;
+      email = credentials[subEmailClaimKey];
+      // If the email wasn't included in the accessToken, try to fetch it from Auth0
+      if (!validateEmail(email)) {
+        email = await this.tryGetEmailFromAuth0(request.auth.token);
+      }
+      if (validateEmail(email)) {
+        valid = true;
+      }
+    } catch (error) {
+      this.logger.error(`Unable to fetch auth0 userinfo`, error);
+    }
+
     // NOTE: we have just a single team at this point so no further checks are necessary
-    callback(undefined, true, payload);
+    callback(undefined, valid, {...payload, [subEmailClaimKey]: email});
+  }
+
+  public async validateAdmin(
+    payload: AccessToken,
+    request: any,
+    callback: (err: any, valid: boolean, credentials?: any) => void,
+  ) {
+    let validEmail = false;
+    let validTeam = false;
+    let email: string | undefined;
+    try {
+      const credentials = payload;
+      email = credentials[subEmailClaimKey];
+      // If the email wasn't included in the accessToken, try to fetch it from Auth0
+      if (!validateEmail(email)) {
+        email = await this.tryGetEmailFromAuth0(request.auth.token);
+      }
+      if (validateEmail(email)) {
+        validEmail = true;
+        const user = await this.gitlab.getUserByEmail(email || '');
```

Comment:
Nope, since email is `string | undefined`. But I realized a much cleaner approach is to add a *type predicate* `email is string` to `validateEmail`.